File: ios/App/Binding/AccountBinding.swift
//
//  This file is part of Blokada.
//
//  This Source Code Form is subject to the terms of the Mozilla Public
//  License, v. 2.0. If a copy of the MPL was not distributed with this
//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
//  Copyright © 2023 Blocka AB. All rights reserved.
//
//  @author Kar
//

import Foundation
import Factory
import Combine

struct AccountWithKeypair {
    let account: JsonAccount
    let keypair: PlusKeypair?
}

struct JsonAccount: Codable {
    let id: String
    let active_until: String?
    let active: Bool?
    let type: String?

    func activeUntil() -> Date? {
        guard let active = active_until else {
            return nil
        }

        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = blockaDateFormat
        guard let date = dateFormatter.date(from: active) else {
            dateFormatter.dateFormat = blockaDateFormatNoNanos
            guard let date = dateFormatter.date(from: active) else {
                return nil
            }
            return date
        }
        return date
    }

    func isActive() -> Bool {
        return active ?? false
    }
}

extension Account: Equatable {
    static func == (lhs: Account, rhs: Account) -> Bool {
        return
            lhs.id == rhs.id &&
            lhs.activeUntil == rhs.activeUntil
    }
}

File: android/app/src/main/java/binding/AccountBinding.kt
/*
 * This file is part of Blokada.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2023 Blocka AB. All rights reserved.
 *
 * @author Karol Gusak (karol@blocka.net)
 */

package binding

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.distinctUntilChanged
import androidx.lifecycle.map
import channel.account.Account
import channel.account.AccountOps
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch
import model.toAccountType
import service.FlutterService
import utils.toBlockaDate
import java.util.Date

fun Account?.getType() = this?.type.toAccountType()
fun Account.activeUntil(): Date = activeUntil?.toBlockaDate() ?: Date(0)

object AccountBinding: AccountOps {
    val account = MutableStateFlow<Account?>(null)
    val live = MutableLiveData<Account>()
    val expiration: LiveData<Date> = live.map { it.activeUntil() }.distinctUntilChanged()

    private val flutter by lazy { FlutterService }
    private val scope = GlobalScope

    init {
        AccountOps.setUp(flutter.engine.dartExecutor.binaryMessenger, this)
        scope.launch {
            account.collect { live.postValue(it) }
        }
    }

    override fun doAccountChanged(account: Account, callback: (Result<Unit>) -> Unit) {
        this.account.value = account
        callback(Result.success(Unit))
    }
}